### DISCUSSION TITLE
UnprovisionSmartStartNode (dsk version) throws exceptions

### DISCUSSION THREAD

**Discussion by georgeinva2004** [zwave-js/ZWaveJS.NET#21] [Q&A]
Hi again!

I successfully was able to add a DSK from a real Zwave device!!!!  I can see the DSK when I call GetProvisioningEntries, but when I invoke UnprovisionSmartStartNode(dsk), I get ObjectDisposedException and OperationCanceledException exceptions.  

<edit>  Finally got a stack trace after running it a few times:

Ran it a few more times, and I finally got a stack trace.  I hope this helps...

 at System.Net.WebSockets.WebSocketBase.OutstandingOperationHelper.ThrowIfDisposed() in f:\dd\NDP\fx\src\net\System\Net\WebSockets\WebSocketBase.cs:line 2477
   at System.Net.WebSockets.WebSocketBase.OutstandingOperationHelper.TryStartOperation(CancellationToken userCancellationToken, CancellationToken& linkedCancellationToken) in f:\dd\NDP\fx\src\net\System\Net\WebSockets\WebSocketBase.cs:line 2351
   at System.Net.WebSockets.WebSocketBase.<CloseOutputAsyncCore>d__51.MoveNext() in f:\dd\NDP\fx\src\net\System\Net\WebSockets\WebSocketBase.cs:line 635

Any ideas?

Thanks!!!

**Comment by marcus-j-davies** [comment]
Hi @georgeinva2004.

I believe this is the `WatsonWebsocket` implementation I am using.

https://github.com/zwave-js/ZWaveJS.NET/blob/main/Visual%20Studio%20Projects/ZWaveJS.NET/ZWaveJS.NET/WSClientClass/WatsonWsClient.cs

https://github.com/zwave-js/ZWaveJS.NET/blob/23318a7ccd40fb14f9bd9c3a690d586f4436628b/Visual%20Studio%20Projects/ZWaveJS.NET/ZWaveJS.NET/WSClientClass/WatsonWsClient.cs#L557

I'm confused why this is only triggered for `UnprovisionSmartStartNode` however.
are you supplying the entire DSK `xxxxx-xxxxx-xxxxx-xxxxx-xxxxx-xxxxx-xxxxx-xxxxx` ?

The WS client does need some improvements - i.e I need to subscribe to certain events to restart it
(and I believe this is why you receive the disposed error)

I'm currently gearing towards 3.1.0 - so will take a look as to why this isn't getting handled.


Any help is appreciated 😉

**Reply by marcus-j-davies** [reply]
Just a thought. Make sure you have provided 32 character (hexadecimal) values for all keys, but to be fair it should be captured by the driver, and passed as an error through `CMDResult` but you never know.

Key checks have been implemented in my 3.1.0 branch - and will gracefully return false via `CMDResult` if they are missing.

**Reply by georgeinva2004** [reply]
Yes, I am giving the entire DSK. I take the actual dsk from GetProvisioningEntries CMDResult and pass it to the Unprovision call. Something interesting just happened.  My machine restarted for updates and I ran it again.  This time, the GetProvisioningEntries() returned a successful status, but no provisioning entries.  I wonder if the entry actually cleared in a prior call, but not reflected in the GetProvisioningEntries() call until I did a complete restart. 

What do you need in the way of help?  I was thinking that a graceful method of shutting down or restarting the server might be helpful.  I noticed that the server doesn't always shutdown when the application is terminated.   I don't have any experience with the WS client you are using, but I have used WebsocketSharp in other projects. it can't be that different ( I know famous last words).  I'm happy to take a crack at whatever you need.

**Reply by marcus-j-davies** [reply]
>  I was thinking that a graceful method of shutting down or restarting the server might be helpful

The Driver class has a public method of `Destroy()`, which will bring everything down (including taking down the process)
https://github.com/zwave-js/ZWaveJS.NET/blob/23318a7ccd40fb14f9bd9c3a690d586f4436628b/Visual%20Studio%20Projects/ZWaveJS.NET/ZWaveJS.NET/Driver.cs#L433

I will investigate this in more detail tomorrow. - I think its just me not subscribing to certain events (where I really should be)

**Reply by marcus-j-davies** [reply]
Could you capture a ZWave log also - I can't imagine anything driver side (I use it in other systems)

like here https://github.com/zwave-js/node-red-contrib-zwave-js

But just want a complete picture  Thank you 🙏

**Reply by georgeinva2004** [reply]
Yeah.  I tried Destroy().   The process is still active.  Let me see what I can find tomorrow (kids stuff tonight)

**Comment by marcus-j-davies** [comment]
Hi @georgeinva2004 

So...I have made a number of changes, both with the library and the platform image (assuming you're using Windows?)

I have attached the PSI to this message if so, I needed to attach it, as the node Image is compiled as part of a routine that occurs in GitHub, when I release a new version. 

I have also uploaded 3.1.0.alpha-1 to NuGet
could you try these updates?

I have... 
 - Server process has been updated to gracefully destroy the driver (so it will finish any writes to its internal DB)
 - I now handle WS Disconnects in the library, the library will re-connect after
 - General optimisations to some internals
 - plus various others documented on the Changelog in the 3.1.0 Branch.
 
To re-cap
 - Use the attached PSI
 - Pull down 3.1.0 Alpha 1 from NuGet
 
 You will note, there is a **DeleteAndDownloadPSI** method - but it won't be staying, so I would ignore it.
 
 Your help in tidying up this release is thoroughly appreciated
 

[server.zip](https://github.com/zwave-js/ZWaveJS.NET/files/8860651/server.zip)

**Reply by georgeinva2004** [reply]
Did a quick test.  The server does terminate now.  I'm no longer getting exceptions calling the UnprovisionSmartStartNode().    I'm doing a more thorough test in a bit.  I'll work through the changelog and do regression testing based on what I find there.  Great work!!!!!

**Reply by marcus-j-davies** [reply]
Avoid the PSI fetch methods - these will fail, until I release 3.1.0 on Github.
All the PSI Binaries are compiled for Linux (+ arm), Win, and MacOS during a release.

The attached PSI above (Win) should be what is delivered for 3.1.0.

the fetch methods will pull them from Github - but 3.1.0 locks the versions to the correct  binary - so they don't yet exist in Github

**Reply by georgeinva2004** [reply]
Did some additional testing.  Looking good so far.  I do see an issue with Smart Start.  I pass the DSK, and I see multiple Zwave Nodes being created with different Node IDs.   When I delete the DSK, the device creation stops.  I verified that the nodes are in my Z-Stick's internal DB by using Silicon Graphics ZWave PC Controller.  This could be my lack of understanding of how Smart Start works.

**Reply by marcus-j-davies** [reply]
Hi @georgeinva2004 

Great - I will continue to work on 3.1.0 - let me know of issues with the Alpha 👍 



> and I see multiple Zwave Nodes being created with different Node IDs

If the same device is being added multiple times, the device maybe failing to get added fully.

Smart Start works like this (I might not be 100% on the Mark)
```
-> A smart start device (that is not joined to a network)
   -> Advertises its self as A Smart Start capable device (this includes advertising its DSK - or partial of)
       -> If the DSK (extracted from the Qr code) matches one we have scanned
            -> Start Inclusion
                 -> From hear, normal inclusion handshake routine (no difference in traditional methods)
```
                 
A device being added multiple times, is usually network related, or some odd device quirk.

if it continues to happen, this might warrant a bug report on the repo for the driver (Debug level driver logs will be needed)

@AlCalzone  - any initial thought.

**Reply by georgeinva2004** [reply]
FYI, the device is a Zooz water leak sensor.

**Reply by AlCalzone** [reply]
This might be worth reporting to Zooz. I've seen reports that some devices crash and reset during the initial interview, restarting the inclusion from scratch.
Driver logs will be good anyways.
